mapviz_plugins/src/gps_plugin.cpp

```diff
@@ -0,0 +1,486 @@
+    p3.setColor(QPalette::Text, Qt::red);
+    ui_.status->setPalette(p3);
+
+    QObject::connect(ui_.selectcolor, SIGNAL(clicked()), this, SLOT(SelectColor()));
+    QObject::connect(ui_.selecttopic, SIGNAL(clicked()), this, SLOT(SelectTopic()));
+    QObject::connect(ui_.topic, SIGNAL(editingFinished()), this, SLOT(TopicEdited()));
+    QObject::connect(ui_.positiontolerance, SIGNAL(valueChanged(double)), this, SLOT(PositionToleranceChanged(double)));
+    QObject::connect(ui_.buffersize, SIGNAL(valueChanged(int)), this, SLOT(BufferSizeChanged(int)));
+    QObject::connect(ui_.drawstyle, SIGNAL(activated(QString)), this, SLOT(SetDrawStyle(QString)));
+  }
+
+  GpsPlugin::~GpsPlugin()
+  {
+  }
+
+  void GpsPlugin::SetDrawStyle(QString style)
+  {
+    if (style == "lines")
+    {
+      draw_style_ = LINES;
+    }
+    else if (style == "points")
+    {
+      draw_style_ = POINTS;
+    }
+    else if (style == "arrows")
+    {
+      draw_style_ = ARROWS;
+    }
+
+    canvas_->update();
+  }
+
+  void GpsPlugin::SelectTopic()
+  {
+    QDialog dialog;
+    Ui::topicselect ui;
+    ui.setupUi(&dialog);
+
+    std::vector<ros::master::TopicInfo> topics;
+    ros::master::getTopics(topics);
+
+    for (unsigned int i = 0; i < topics.size(); i++)
+    {
+      if (topics[i].datatype == "gps_common/GPSFix")
+      {
+        ui.displaylist->addItem(topics[i].name.c_str());
+      }
+    }
+    ui.displaylist->setCurrentRow(0);
+
+    dialog.exec();
+
+    if (dialog.result() == QDialog::Accepted && ui.displaylist->selectedItems().count() == 1)
+    {
+      ui_.topic->setText(ui.displaylist->selectedItems().first()->text());
+      TopicEdited();
+    }
+  }
+
+  void GpsPlugin::SelectColor()
+  {
+    QColorDialog dialog(color_, config_widget_);
+    dialog.exec();
+
+    if (dialog.result() == QDialog::Accepted)
+    {
+      color_ = dialog.selectedColor();
+      ui_.selectcolor->setStyleSheet("background: " + color_.name() + ";");
+      canvas_->update();
+    }
+  }
+
+  void GpsPlugin::TopicEdited()
+  {
+    if (ui_.topic->text().toStdString() != topic_)
+    {
+      initialized_ = false;
+      points_.clear();
+      has_message_ = false;
+      topic_ = ui_.topic->text().toStdString();
+      PrintWarning("No messages received.");
+
+      gps_sub_.shutdown();
+      gps_sub_ = node_.subscribe(topic_, 1, &GpsPlugin::GPSFixCallback, this);
+
+      ROS_INFO("Subscribing to %s", topic_.c_str());
+    }
+  }
+
+  void GpsPlugin::GPSFixCallback(const gps_common::GPSFixConstPtr gps)
+  {
+    if (!local_xy_util_.Initialized())
+    {
+      return;
+    }
+    if (!has_message_)
+    {
+      initialized_ = true;
+      has_message_ = true;
+    }
+
+    StampedPoint stamped_point;
+    stamped_point.stamp = gps->header.stamp;
+
+    double x;
+    double y;
+    local_xy_util_.ToLocalXy(gps->latitude, gps->longitude, x, y);
+
+    stamped_point.point = tf::Point(x, y, gps->altitude);
+
+    stamped_point.orientation = tf::Quaternion(
```

Comment:
The GPS track value can be used here to get the orientation so the arrow view will display correctly.